File: DL-ATM41G2/DL-ATM41G2.ex

# https://www.decentlab.com/products/eleven-parameter-weather-station-for-lorawan

defmodule DecentlabDecoder do
  @protocol_version 2

  defp sensor_defs do
    [
      %{
        length: 17,
        values: [
          %{
            :name => "Solar radiation",
            :convert => fn x -> (Enum.at(x, 0) - 32768) / 10 end,
            :unit => "W⋅m⁻²"
          },
          %{
            :name => "Precipitation",
            :convert => fn x -> Enum.at(x, 1) / 1000 end,
            :unit => "mm"
          },
          %{
            :name => "Lightning strike count",
            :convert => fn x -> Enum.at(x, 2) - 32768 end,
            :unit => "None"
          },
          %{
            :name => "Lightning average distance",
            :convert => fn x -> Enum.at(x, 3) - 32768 end,
            :unit => "km"
          },
          %{
            :name => "Wind speed",
            :convert => fn x -> (Enum.at(x, 4) - 32768) / 100 end,
            :unit => "m⋅s⁻¹"
          },
          %{
            :name => "Wind direction",
            :convert => fn x -> (Enum.at(x, 5) - 32768) / 10 end,
            :unit => "°"
          },
          %{
            :name => "Maximum wind speed",
            :convert => fn x -> (Enum.at(x, 6) - 32768) / 100 end,
            :unit => "m⋅s⁻¹"
          },
          %{
            :name => "Air temperature",
            :convert => fn x -> (Enum.at(x, 7) - 32768) / 10 end,
            :unit => "°C"
          },
          %{
            :name => "Vapor pressure",
            :convert => fn x -> (Enum.at(x, 8) - 32768) / 100 end,
            :unit => "kPa"
          },
          %{
            :name => "Barometric pressure",
            :convert => fn x -> (Enum.at(x, 9) - 32768) / 100 end,
            :unit => "kPa"
          },
          %{
            :name => "Relative humidity",
            :convert => fn x -> (Enum.at(x, 10) - 32768) / 10 end,
            :unit => "%"
          },
          %{
            :name => "Internal temperature",
            :convert => fn x -> (Enum.at(x, 11) - 32768) / 10 end,
            :unit => "°C"
          },
          %{
            :name => "Tilt angle, X orientation",
            :convert => fn x -> (Enum.at(x, 12) - 32768) / 10 end,
            :unit => "°"
          },
          %{
            :name => "Tilt angle, Y orientation",
            :convert => fn x -> (Enum.at(x, 13) - 32768) / 10 end,
            :unit => "°"
          },
          %{
            :name => "Precipitation electrical conductivity",
            :convert => fn x -> Enum.at(x, 14) - 32768 end,
            :unit => "µS⋅cm⁻¹"
          },
          %{
            :name => "Cumulative precipitation",
            :convert => fn x -> (Enum.at(x, 15) + Enum.at(x, 16) * 65536) / 1000 end,
            :unit => "mm"
          }
        ]
      },
      %{
        length: 1,
        values: [
          %{
            :name => "Battery voltage",
            :convert => fn x -> Enum.at(x, 0) / 1000 end,
            :unit => "V"
          }
        ]
      }
    ]
  end


  def decode(msg, :hex) do
    {:ok, bytes} = Base.decode16(msg, case: :mixed)
    decode(bytes)
  end

  def decode(msg) when is_binary(msg), do: decode_binary(msg)

  def decode(msg), do: to_string(msg) |> decode

  defp decode_binary(<<@protocol_version, device_id::size(16), flags::binary-size(2), bytes::binary>>) do
    bytes
    |> bytes_to_words()
    |> sensor(flags, sensor_defs())
    |> Map.put("Device ID", device_id)
    |> Map.put("Protocol version", @protocol_version)
  end


  defp bytes_to_words(<<>>), do: []

  defp bytes_to_words(<<word::size(16), rest::binary>>), do: [word | bytes_to_words(rest)]


  defp sensor(words, <<flags::size(15), 1::size(1)>>, [%{length: len, values: value_defs} | rest]) do
    {x, rest_words} = Enum.split(words, len)
    value(value_defs, x)
    |> Map.merge(sensor(rest_words, <<0::size(1), flags::size(15)>>, rest))
  end

  defp sensor(words, <<flags::size(15), 0::size(1)>>, [_cur | rest]) do
    sensor(words, <<0::size(1), flags::size(15)>>, rest)
  end

  defp sensor([], _flags, []), do: %{}


  defp value([], _x), do: %{}

  defp value([%{convert: nil} | rest], x), do: value(rest, x)

  defp value([%{name: name, unit: unit, convert: convert} | rest], x) do
    value(rest, x)
    |> Map.put(name, %{"unit" => unit, "value" => convert.(x)})
  end

  defp where(true, if_true, _if_false), do: if_true
  defp where(false, _if_true, if_false), do: if_false

end

IO.inspect(DecentlabDecoder.decode("025ef80003805c000080008000803484b3803680e78086a60181d680ed81c9809f8000117000010adc", :hex))
IO.inspect(DecentlabDecoder.decode("025ef800020adc", :hex))

File: DL-ATM41G2/DL-ATM41G2.erl

% https://www.decentlab.com/products/eleven-parameter-weather-station-for-lorawan

-module(decentlab_decoder).
-define(PROTOCOL_VERSION, 2).
-export([decode/1, start/0]).

-define(SENSOR_DEFS,
  [
    #{
      length => 17,
      values => [
        #{
          name => <<"Solar radiation">>,
          convert => fun(X) -> (lists:nth(0 + 1, X) - 32768) / 10 end,
          unit => <<"W⋅m⁻²"/utf8>>
        },
        #{
          name => <<"Precipitation">>,
          convert => fun(X) -> lists:nth(1 + 1, X) / 1000 end,
          unit => <<"mm"/utf8>>
        },
        #{
          name => <<"Lightning strike count">>,
          convert => fun(X) -> lists:nth(2 + 1, X) - 32768 end,
          unit => <<"None"/utf8>>
        },
        #{
          name => <<"Lightning average distance">>,
          convert => fun(X) -> lists:nth(3 + 1, X) - 32768 end,
          unit => <<"km"/utf8>>
        },
        #{
          name => <<"Wind speed">>,
          convert => fun(X) -> (lists:nth(4 + 1, X) - 32768) / 100 end,
          unit => <<"m⋅s⁻¹"/utf8>>
        },
        #{
          name => <<"Wind direction">>,
          convert => fun(X) -> (lists:nth(5 + 1, X) - 32768) / 10 end,
          unit => <<"°"/utf8>>
        },
        #{
          name => <<"Maximum wind speed">>,
          convert => fun(X) -> (lists:nth(6 + 1, X) - 32768) / 100 end,
          unit => <<"m⋅s⁻¹"/utf8>>
        },
        #{
          name => <<"Air temperature">>,
          convert => fun(X) -> (lists:nth(7 + 1, X) - 32768) / 10 end,
          unit => <<"°C"/utf8>>
        },
        #{
          name => <<"Vapor pressure">>,
          convert => fun(X) -> (lists:nth(8 + 1, X) - 32768) / 100 end,
          unit => <<"kPa"/utf8>>
        },
        #{
          name => <<"Barometric pressure">>,
          convert => fun(X) -> (lists:nth(9 + 1, X) - 32768) / 100 end,
          unit => <<"kPa"/utf8>>
        },
        #{
          name => <<"Relative humidity">>,
          convert => fun(X) -> (lists:nth(10 + 1, X) - 32768) / 10 end,
          unit => <<"%"/utf8>>
        },
        #{
          name => <<"Internal temperature">>,
          convert => fun(X) -> (lists:nth(11 + 1, X) - 32768) / 10 end,
          unit => <<"°C"/utf8>>
        },
        #{
          name => <<"Tilt angle, X orientation">>,
          convert => fun(X) -> (lists:nth(12 + 1, X) - 32768) / 10 end,
          unit => <<"°"/utf8>>
        },
        #{
          name => <<"Tilt angle, Y orientation">>,
          convert => fun(X) -> (lists:nth(13 + 1, X) - 32768) / 10 end,
          unit => <<"°"/utf8>>
        },
        #{
          name => <<"Precipitation electrical conductivity">>,
          convert => fun(X) -> lists:nth(14 + 1, X) - 32768 end,
          unit => <<"µS⋅cm⁻¹"/utf8>>
        },
        #{
          name => <<"Cumulative precipitation">>,
          convert => fun(X) -> (lists:nth(15 + 1, X) + lists:nth(16 + 1, X) * 65536) / 1000 end,
          unit => <<"mm"/utf8>>
        }
      ]
    },
    #{
      length => 1,
      values => [
        #{
          name => <<"Battery voltage">>,
          convert => fun(X) -> lists:nth(0 + 1, X) / 1000 end,
          unit => <<"V"/utf8>>
        }
      ]
    }
  ]
).


decode(<<?PROTOCOL_VERSION, DeviceId:16,
         Flags:16/bitstring, Bytes/binary>> = Binary) when is_binary(Binary) ->
  Words = bytes_to_words(Bytes),
  maps:merge(#{<<"Protocol version">> => ?PROTOCOL_VERSION,
               <<"Device ID">> => DeviceId},
             sensor(Words, Flags, ?SENSOR_DEFS));
decode(HexStr) ->
  Binary = hex2bin(HexStr),
  decode(Binary).


start() ->
  io:format("~p~n", [decode("025ef80003805c000080008000803484b3803680e78086a60181d680ed81c9809f8000117000010adc")]),
  io:format("~p~n", [decode("025ef800020adc")]).


%%%%%%%%%%%%%%%%%%%%%%%%%%%%% PRIVATE FUNCTIONS %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

sensor(_Words, _Flags, []) -> #{};
sensor(Words, <<Flags:15, 0:1>>, [_Cur | Rest]) ->
  sensor(Words, <<0:1, Flags:15>>, Rest);
sensor(Words, <<Flags:15, 1:1>>,
       [#{length := Len, values := ValueDefs} | RestDefs]) ->
  {X, RestWords} = lists:split(Len, Words),
  Values = value(ValueDefs, X),
  RestSensors = sensor(RestWords, <<0:1, Flags:15>>, RestDefs),
  maps:merge(Values, RestSensors).


value([], _X) -> #{};
value([#{convert := Convert,
         name := Name} = ValueDef | RestDefs], X) ->
  Values = value(RestDefs, X),
  Value = #{<<"value">> => Convert(X)},
  Value2 = add_unit(Value, ValueDef),
  maps:put(Name, Value2, Values);
value([_NoConvert | RestDefs], X) -> value(RestDefs, X).


add_unit(Value, #{unit := Unit}) ->
  maps:put(<<"unit">>, Unit, Value);
add_unit(Value, _NoUnit) ->
  Value.


hex2bin(HexStr) when is_binary(HexStr) -> hex2bin(binary_to_list(HexStr));
hex2bin(HexStr) -> hex2bin(HexStr, []).

hex2bin("", Bins) -> list_to_binary(lists:reverse(Bins));
hex2bin(HexStr, Bins) ->
  Bytes = string:slice(HexStr, 0, 2),
  Word = list_to_integer(Bytes, 16),
  hex2bin(string:slice(HexStr, 2), [Word | Bins]).


bytes_to_words(Bytes) -> bytes_to_words(Bytes, []).

bytes_to_words(<<>>, Words) -> lists:reverse(Words);
bytes_to_words(<<Word:16, Rest/binary>>, Words) ->
  bytes_to_words(Rest, [Word | Words]).

where(true, IfTrue, _IfFalse) -> IfTrue;
where(false, _IfTrue, IfFalse) -> IfFalse.